File: 강의 자료/02-알고리즘/12-트리/문제별 코드/4803-트리/solution.py
import sys
from collections import deque
si = sys.stdin.readline
def solve(tt):
    n, m = map(int, si().split())
    if n == 0 and m == 0:
        return False
    adj = [[] for _ in range(n + 1)]
    visit = [0] * (n + 1)
    sys.setrecursionlimit(100005)

    for _ in range(m):
        x, y = list(map(int, si().split()))
        adj[x].append(y)
        adj[y].append(x)

    vertex_cnt = 0
    edge_cnt = 0
    def dfs(x):
        nonlocal vertex_cnt, edge_cnt
        visit[x] = True
        vertex_cnt += 1
        edge_cnt += len(adj[x])
        for y in adj[x]:
            if visit[y]:
                continue
            dfs(y)

    ans = 0
    for i in range(1, n + 1):
        if visit[i]: continue
        vertex_cnt, edge_cnt = 0, 0
        dfs(i)
        if edge_cnt == (vertex_cnt - 1) * 2:
            ans += 1
    print(f'Case {tt}: ', end='')
    if ans == 0:
        print("No trees.")
    elif ans == 1:
        print("There is one tree.")
    else:
        print(f"A forest of {ans} trees.")
    return True

tt = 0
while True:
    tt = tt + 1
    if not solve(tt):
        break

File: 강의 자료/02-알고리즘/12-트리/문제별 코드/4803-트리/solution.java
import java.io.*;
import java.lang.reflect.Array;
import java.util.*;

public class Main {
    static FastReader scan = new FastReader();
    static StringBuilder sb = new StringBuilder();

    static int n, m, vertex_cnt, edge_cnt;
    static ArrayList<Integer>[] adj;
    static boolean[] visit;

    static void input() {
        n = scan.nextInt();
        m = scan.nextInt();
        adj = new ArrayList[n + 1];
        visit = new boolean[n + 1];
        for (int i = 1; i <= n; i++) adj[i] = new ArrayList<>();

        // 인접 리스트 구성하기
        for (int i = 1; i <= m; i++) {
            int x = scan.nextInt(), y = scan.nextInt();
            adj[x].add(y);
            adj[y].add(x);
        }
    }

    static void dfs(int x) {
        vertex_cnt++;
        edge_cnt += adj[x].size();
        visit[x] = true;
        for (int y : adj[x]) {
            if (visit[y]) continue;
            dfs(y);
        }
    }

    static void pro(int tt) {
        int ans = 0;
        for (int i = 1; i <= n; i++){
            if (visit[i]) continue;
            vertex_cnt = 0;
            edge_cnt = 0;
            dfs(i);
            if (edge_cnt == (vertex_cnt - 1) * 2) ans++;
        }

        // 정답 출력하기
        sb.append("Case ").append(tt).append(": ");
        if (ans == 0){
            sb.append("No trees.\n");
        } else if (ans == 1){
            sb.append("There is one tree.\n");
        } else {
            sb.append("A forest of ").append(ans).append(" trees.\n");
        }
    }

    public static void main(String[] args) {
        for (int tt = 1;;tt++){
            input();
            if (n == 0 && m == 0) break;
            pro(tt);
        }
        System.out.print(sb);
    }


    static class FastReader {
        BufferedReader br;
        StringTokenizer st;

        public FastReader() {
            br = new BufferedReader(new InputStreamReader(System.in));
        }

        public FastReader(String s) throws FileNotFoundException {
            br = new BufferedReader(new FileReader(new File(s)));
        }

        String next() {
            while (st == null || !st.hasMoreElements()) {
                try {
                    st = new StringTokenizer(br.readLine());
                } catch (IOException e) {
                    e.printStackTrace();
                }
            }
            return st.nextToken();
        }

        int nextInt() {
            return Integer.parseInt(next());
        }

        long nextLong() {
            return Long.parseLong(next());
        }

        double nextDouble() {
            return Double.parseDouble(next());
        }

        String nextLine() {
            String str = "";
            try {
                str = br.readLine();
            } catch (IOException e) {
                e.printStackTrace();
            }
            return str;
        }
    }
}